File: sage/services/snarc/detectors/surprise.py
"""
Surprise Detector - Prediction Error

Detects deviation from expected sensor values.
Learns to predict next sensor state, measures surprise as prediction error.
"""

import time as _time

import torch
import numpy as np
from typing import Dict, Any, Optional
from collections import deque

from sage.services.snarc.temporal import TimestampedDeque, DEFAULT_HALF_LIVES


class SimplePredictorEMA:
    """
    Simple Exponential Moving Average predictor

    Predicts next value as weighted average of past observations.
    Good baseline for smooth time series.
    """

    def __init__(self, alpha: float = 0.3):
        """
        Args:
            alpha: Smoothing factor (0-1). Higher = more weight to recent.
        """
        self.alpha = alpha
        self.prediction: Optional[Any] = None

    def predict(self) -> Optional[Any]:
        """Get current prediction"""
        return self.prediction

    def update(self, observation: Any):
        """Update predictor with new observation"""
        if self.prediction is None:
            # First observation - just store it
            self.prediction = observation
        else:
            # EMA update
            if isinstance(observation, torch.Tensor):
                self.prediction = (
                    self.alpha * observation +
                    (1 - self.alpha) * self.prediction
                )
            elif isinstance(observation, np.ndarray):
                self.prediction = (
                    self.alpha * observation +
                    (1 - self.alpha) * self.prediction
                )
            elif isinstance(observation, (int, float)):
                self.prediction = (
                    self.alpha * observation +
                    (1 - self.alpha) * self.prediction
                )
            else:
                # Unknown type - just store
                self.prediction = observation


class SurpriseDetector:
    """
    Detect deviation from prediction

    Maintains predictor per sensor, computes surprise as
    normalized distance between prediction and actual.
    """

    def __init__(self, alpha: float = 0.3, history_size: int = 100,
                 half_life: float = DEFAULT_HALF_LIVES['surprise']):
        """
        Args:
            alpha: Smoothing factor for EMA predictor
            history_size: How many past surprise scores to keep
            half_life: Time-decay half-life in seconds for normalization weighting
        """
        self.alpha = alpha
        self.history_size = history_size
        self.half_life = half_life

        # Predictor per sensor
        self.predictors: Dict[str, SimplePredictorEMA] = {}

        # Surprise history per sensor (for normalization) — original deque kept for compat
        self.surprise_history: Dict[str, deque] = {}
        # Time-aware history for weighted normalization
        self._timed_history: Dict[str, TimestampedDeque] = {}

    def compute(self, sensor_output: Any, sensor_id: str) -> float:
        """
        Compute surprise for sensor observation

        Args:
            sensor_output: Current sensor reading
            sensor_id: Unique sensor identifier

        Returns:
            Surprise score (0.0-1.0)
                0.0 = perfectly predicted
                1.0 = maximally surprising
        """
        # Get or create predictor for this sensor
        if sensor_id not in self.predictors:
            self.predictors[sensor_id] = SimplePredictorEMA(self.alpha)
            self.surprise_history[sensor_id] = deque(maxlen=self.history_size)
            self._timed_history[sensor_id] = TimestampedDeque(maxlen=self.history_size)

        predictor = self.predictors[sensor_id]
        prediction = predictor.predict()

        # If first observation, no surprise yet
        if prediction is None:
            predictor.update(sensor_output)
            return 0.0

        # Compute prediction error
        error = self._compute_distance(sensor_output, prediction)

        # Normalize using history
        normalized_surprise = self._normalize_surprise(error, sensor_id)

        # Update predictor and history
        now = _time.time()
        predictor.update(sensor_output)
        self.surprise_history[sensor_id].append(error)
        self._timed_history[sensor_id].append(error, now)

        return normalized_surprise

    def _compute_distance(self, actual: Any, predicted: Any) -> float:
        """
        Compute distance between actual and predicted

        Handles different data types (tensors, arrays, scalars)
        """
        if isinstance(actual, torch.Tensor) and isinstance(predicted, torch.Tensor):
            # L2 distance for tensors
            dist = torch.norm(actual - predicted).item()
            return dist

        elif isinstance(actual, np.ndarray) and isinstance(predicted, np.ndarray):
            # L2 distance for arrays
            dist = np.linalg.norm(actual - predicted)
            return float(dist)

        elif isinstance(actual, (int, float)) and isinstance(predicted, (int, float)):
            # Absolute difference for scalars
            return abs(actual - predicted)

        else:
            # Unknown types - no meaningful distance
            return 0.0

    def _normalize_surprise(self, error: float, sensor_id: str) -> float:
        """
        Normalize surprise using historical distribution with time-decay weighting.

        Recent observations contribute more to the normalization distribution
        than old ones. Uses percentile-based normalization:
        - If error is at 50th percentile → surprise = 0.5
        - If error is at 95th percentile → surprise = 0.95
        """
        timed = self._timed_history.get(sensor_id)

        if timed is None or len(timed) < 10:
            return min(error / 10.0, 1.0)

        return timed.weighted_percentile_of(error, self.half_life)

    def reset_sensor(self, sensor_id: str):
        """Reset predictor for specific sensor"""
        if sensor_id in self.predictors:
            del self.predictors[sensor_id]
        if sensor_id in self.surprise_history:
            del self.surprise_history[sensor_id]
        if sensor_id in self._timed_history:
            del self._timed_history[sensor_id]

    def reset_all(self):
        """Reset all predictors"""
        self.predictors.clear()
        self.surprise_history.clear()
        self._timed_history.clear()

File: sage-rs/sage-lib/src/snarc/surprise.rs
use std::collections::HashMap;

use crate::snarc::temporal::{default_half_life, TimestampedDeque, now_secs};

/// Simple Exponential Moving Average predictor.
/// Predicts next value as weighted average of past observations.
pub struct SimplePredictorEMA {
    alpha: f64,
    prediction: Option<f64>,
}

impl SimplePredictorEMA {
    pub fn new(alpha: f64) -> Self {
        Self {
            alpha,
            prediction: None,
        }
    }

    pub fn predict(&self) -> Option<f64> {
        self.prediction
    }

    pub fn update(&mut self, observation: f64) {
        self.prediction = Some(match self.prediction {
            None => observation,
            Some(prev) => self.alpha * observation + (1.0 - self.alpha) * prev,
        });
    }
}

/// Surprise detector — measures prediction error against an EMA baseline,
/// normalizes via time-weighted percentile rank over recent history.
pub struct SurpriseDetector {
    alpha: f64,
    history_size: usize,
    half_life: f64,
    predictors: HashMap<String, SimplePredictorEMA>,
    timed_history: HashMap<String, TimestampedDeque>,
}

impl SurpriseDetector {
    pub fn new(alpha: f64, history_size: usize, half_life: Option<f64>) -> Self {
        Self {
            alpha,
            history_size,
            half_life: half_life.unwrap_or_else(|| default_half_life("surprise")),
            predictors: HashMap::new(),
            timed_history: HashMap::new(),
        }
    }

    pub fn with_defaults() -> Self {
        Self::new(0.3, 100, None)
    }

    /// Compute surprise for a scalar observation on the given sensor.
    /// Returns a score in [0.0, 1.0].
    pub fn compute(&mut self, observation: f64, sensor_id: &str) -> f64 {
        if !self.predictors.contains_key(sensor_id) {
            self.predictors
                .insert(sensor_id.to_string(), SimplePredictorEMA::new(self.alpha));
            self.timed_history.insert(
                sensor_id.to_string(),
                TimestampedDeque::new(self.history_size),
            );
        }

        let prediction = self.predictors.get(sensor_id).unwrap().predict();

        if prediction.is_none() {
            self.predictors.get_mut(sensor_id).unwrap().update(observation);
            return 0.0;
        }

        let predicted = prediction.unwrap();
        let error = (observation - predicted).abs();
        let normalized = self.normalize_surprise(error, sensor_id);

        let now = now_secs();
        self.predictors.get_mut(sensor_id).unwrap().update(observation);
        self.timed_history
            .get_mut(sensor_id)
            .unwrap()
            .append(error, Some(now));

        normalized
    }

    fn normalize_surprise(&self, error: f64, sensor_id: &str) -> f64 {
        let timed = match self.timed_history.get(sensor_id) {
            Some(t) if t.len() >= 10 => t,
            _ => return (error / 10.0).min(1.0),
        };

        timed.weighted_percentile_of(error, self.half_life, None)
    }

    pub fn reset_sensor(&mut self, sensor_id: &str) {
        self.predictors.remove(sensor_id);
        self.timed_history.remove(sensor_id);
    }

    pub fn reset_all(&mut self) {
        self.predictors.clear();
        self.timed_history.clear();
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ema_first_observation() {
        let mut ema = SimplePredictorEMA::new(0.3);
        assert!(ema.predict().is_none());
        ema.update(5.0);
        assert_eq!(ema.predict(), Some(5.0));
    }

    #[test]
    fn ema_converges_toward_constant() {
        let mut ema = SimplePredictorEMA::new(0.3);
        for _ in 0..100 {
            ema.update(10.0);
        }
        assert!((ema.predict().unwrap() - 10.0).abs() < 1e-10);
    }

    #[test]
    fn ema_tracks_step_change() {
        let mut ema = SimplePredictorEMA::new(0.3);
        for _ in 0..50 {
            ema.update(1.0);
        }
        // Step to 10.0
        for _ in 0..50 {
            ema.update(10.0);
        }
        // Should be very close to 10.0 after 50 updates at alpha=0.3
        assert!((ema.predict().unwrap() - 10.0).abs() < 0.01);
    }

    #[test]
    fn surprise_first_observation_is_zero() {
        let mut det = SurpriseDetector::with_defaults();
        assert_eq!(det.compute(5.0, "test"), 0.0);
    }

    #[test]
    fn surprise_constant_stream_low() {
        let mut det = SurpriseDetector::with_defaults();
        // Feed 20 identical observations to build history
        for _ in 0..20 {
            det.compute(5.0, "test");
        }
        // Next identical observation should be low surprise
        let s = det.compute(5.0, "test");
        assert!(s < 0.1, "constant stream surprise={s}, expected <0.1");
    }

    #[test]
    fn surprise_spike_is_high() {
        let mut det = SurpriseDetector::with_defaults();
        // Build baseline
        for _ in 0..30 {
            det.compute(1.0, "test");
        }
        // Big spike
        let s = det.compute(100.0, "test");
        assert!(s > 0.8, "spike surprise={s}, expected >0.8");
    }

    #[test]
    fn surprise_independent_sensors() {
        let mut det = SurpriseDetector::with_defaults();
        // First obs on each sensor → 0.0
        assert_eq!(det.compute(5.0, "a"), 0.0);
        assert_eq!(det.compute(100.0, "b"), 0.0);

        // Sensor "a" gets constant, "b" gets spike relative to itself
        for _ in 0..20 {
            det.compute(5.0, "a");
            det.compute(100.0, "b");
        }
        // Both should have low surprise for their respective constants
        let sa = det.compute(5.0, "a");
        let sb = det.compute(100.0, "b");
        assert!(sa < 0.1, "sensor a surprise={sa}");
        assert!(sb < 0.1, "sensor b surprise={sb}");
    }

    #[test]
    fn surprise_reset_sensor() {
        let mut det = SurpriseDetector::with_defaults();
        for _ in 0..20 {
            det.compute(5.0, "test");
        }
        det.reset_sensor("test");
        // After reset, first observation is again zero
        assert_eq!(det.compute(5.0, "test"), 0.0);
    }

    #[test]
    fn surprise_20_observation_parity() {
        // Feed a sequence of 20 scalar observations and verify the detector
        // produces reasonable monotonic surprise for divergent values.
        let mut det = SurpriseDetector::new(0.3, 100, Some(60.0));
        let observations = [
            1.0, 1.1, 0.9, 1.0, 1.05, 0.95, 1.0, 1.02, 0.98, 1.0,
            1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0,
        ];
        let mut scores = Vec::new();
        for &obs in &observations {
            scores.push(det.compute(obs, "parity"));
        }

        // After 20 near-constant observations, a big deviation should spike
        let spike = det.compute(5.0, "parity");
        let last_calm = *scores.last().unwrap();
        assert!(
            spike > last_calm,
            "spike ({spike}) should exceed calm ({last_calm})"
        );
    }
}
